docs/backup-restore-integration.md

```diff
@@ -0,0 +1,334 @@
+
+1.  Populate PersistentVolumeClaims with snapshot data
+2.  Sanitize and apply all relevant Kubernetes resource definitions
+
+Step 1 is beyond the scope of this document.
+
+## Existing KubeVirt Backup Solutions
+
+### Velero Plugin
+
+[Velero](https://velero.io/) is a popular tool for backing up/migrating Kubernetes clusters.  The KubeVirt team actively maintains a [plugin](https://github.com/kubevirt/kubevirt-velero-plugin) for use with Velero.  The plugin implements much of the logic described in this document.
+
+### VirtualMachineSnapshot + VirtualMachineExport API
+
+The [VirtualMachineSnapshot API](https://kubevirt.io/user-guide/operations/snapshot_restore_api/) provides an easy way for KubeVirt users to backup VirtualMachines within a cluster.  On its own, it is not suitable for offsite backup or disaster recovery.  But combined with the [VirtualMachineExport API](https://kubevirt.io/user-guide/operations/export_api/), VirtualMachine volumes may be made available for copy to remote locations.
+
+## Building the KubeVirt Object Graph
+
+![Object Graph Example](backup-graph.png "VM Graph")
+
+In this section, KubeVirt resources and their relationships will be explored by showing how yaml snippets map to nodes in an object graph.
+
+Nodes in the object graph are represented by the following tuple:
+
+(APIGroup, Kind, namespace, name)
+
+### VirtualMachine Object Graph
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm1
+  namespace: ns1
+...
+```
+
+- ("kubevirt.io", "VirtualMachine", "ns1", "vm1")
+
+#### spec.instancetype
+
+```yaml
+...
+spec:
+  instancetype:
+    kind: VirtualMachineInstancetype
+    name: small
+...
+```
+
+- ("kubevirt.io", "VirtualMachineInstancetype", "ns1", "small")
+
+```yaml
+...
+spec:
+  instancetype:
+    kind: VirtualMachineClusterInstancetype
+    name: csmall
+...
+```
+
+- ("kubevirt.io", "VirtualMachineClusterInstancetype", "", "csmall")
+
+#### spec.preference
+
+```yaml
+...
+spec:
+  preference:
+    kind: VirtualMachinePreference
+    name: windows
+...
+```
+
+- ("kubevirt.io", "VirtualMachineClusterPreference", "ns1", "windows")
+
+```yaml
+...
+spec:
+  preference:
+    kind: VirtualMachineClusterPreference
+    name: cwindows
+...
+```
+
+- ("kubevirt.io", "VirtualMachineClusterPreference", "", "cwindows")
+
+#### spec.template
+
+See [VirtualMachineInstance](#virtualmachineinstance-object-graph)
+
+### VirtualMachineInstance Object Graph
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi1
+  namespace: ns1
+...
+```
+
+- ("kubevirt.io", "VirtualMachineInstance", "ns1", "vmi1")
+- ("", "Pod", "ns1", "virt-launcher-vmi1-XXXXX") \*
+
+\* Each VirtualMachineInstance has a corresponding uniquely named Pod.  The backup process can look up the name of this pod by using the `kubevirt.io/created-by=<UID of VirtualMachineInstance>` label selector.
+
+#### spec.priorityClassName
+
+```yaml
+...
+spec:
+  priorityClassName: pc1
```

Comment:
regarding cluster scoped stuff also after talking with Adam it seems like should not be backedup and restored and one should make sure the cluster has the appropriate setting and objects when restoring a vm. those stuff has usually defined configurations that will still exist when restoring to a cluster

Reply:
okay